Update: Changelog for 15.0-beta3 (#14407)
OpenTTD/OpenTTD: changelog.md

- Change: [NewGRF] Animation-trigger 'construction stage changed' of houses and industries now also triggers at construction start. (#14089)
- Change: [Script] Reject scripts using negative version (#14096)
- Change: [NewGRF] Provide shared random bits in multi-tile animation-triggers of airport tiles and objects, just like for other features. (#14090)
- Change: Move Game Options toggles to before label, add Off/On text. (#14097)
- Change: Remove minimum width from advanced settings panel of Game Options window. (#14088)
- Change: Merge Game Options and Game Settings together. (#13242)
- Change: Enhance visibility of disabled bool button. (#14071)
- Change: Use scaled instead of fixed dimensions for framerate graphs. (#14039)
- Change: Include dragged train in depot tile length display. (#14060)
- Change: Use standard boolean toggle buttons for game options toggles. (#14051)
- Change: "setting" console command now shows default value (#14061)
- Change: Allow Double-Ctrl+Click on default size box to clear saved size. (#14055)
- Change: Don't select content when toggling its download status. (#14059)

> - Fix: Scale graph gridlines and axes with GUI scale (#12131)

Reviewer: Should be in PR number order, no?

Reviewer 2: Indeed, reverse PR number order (by type). Also strip out commit hash references

## 15.x

### 15.0-beta3 (2025-07-01)

- Add: Buttons to change picker preview image height. (#14373)
- Add: Show and sort online content version in content list. (#14355)

> - Add: Hover on graph legend to highlight line.

Reviewer: ```suggestion
- Add: Hover on graph legend to highlight line. (#14314)
```

- Add: Allow separate expansion of town buildings and roads in scenario editor. (#14341)

> - Add: [Script] Cloning ScriptList

Reviewer: ```suggestion
- Add: [Script] Cloning ScriptList (#14305)
```

- Add: [Script] Framework for cloning selected ScriptObject

> - Add: Badge dropdown filters and configuration.for picker and vehicle windows.

Reviewer: ```suggestion
- Add: Badge dropdown filters and configuration for picker and vehicle windows. (#14021)
```

> - Add: Configuration of NewGRF badges for picker and vehicle windows.

Reviewer: ```suggestion
- Add: Configuration of NewGRF badges for picker and vehicle windows. (#13993)
```

> - Add: Generic "configure list/view" vertical ellipsis icon.

Reviewer: ```suggestion
```

- Add: [Script] GSBaseStation::GetOwner (#13406)
- Add: [Script] ScriptVehicleList_Waypoint (#13456)

> - Add: [NewGRF] All callbacks returning D0xx strings, have now the option to return any string id via register 0x100.

Reviewer: Maybe too technical?

```suggestion
- Add: [NewGRF] All callbacks returning D0xx strings, have now the option to return any string id via register 0x100. (#14224)
```

> - Add: [NewGRF] Special value 0x7FFE for VarAction2 results specifying 'return calculated result'.
> - Add: [NewGRF] Special value 0x7FFF for Action2 references specifying explicit 'callback/sprite-resolving failed'.

Reviewer: ```suggestion
- Add: [NewGRF] Special value 0x7FFE for VarAction2 results specifying 'return calculated result'. (#14149)
- Add: [NewGRF] Special value 0x7FFF for Action2 references specifying explicit 'callback/sprite-resolving failed'. (#14149)
```

- Add: [NewGRF] Add purchase list Action3 CID for houses, industries, industry tiles, airports and airport tiles. (#14121)
- Add: "Many random towns" button in scenario editor now prompts for the number of towns, with defaults based on new game settings (#14158)

> - Add: [BaseSet] Allow basesets to set minor and patch versions in obg/obs/obm files.

Reviewer: ```suggestion
- Add: [BaseSet] Allow basesets to set minor and patch versions in obg/obs/obm files. (#14169)
```

- Add: Limit height of settings description, and add scrollbar. (#14102)
- Add: [NewGRF] Provide random bits in var10 of house callback 1C 'construction stage changed'. (#14095)
- Add: [NewGRF] Station/roadstop animation-triggers 'tile loop' (bit 7) and 'path reservation' (bit 8). (#14080)
- Change: Improve sprite ellipsis appearance. (#14404)
- Change: New company face definition system and UI. (#14319)

> - Change: Simplify sprite cache memory management

Reviewer: Not player-facing.

```suggestion
```

> - Change: Tweak layout of music player widgets.

Reviewer: Not player-facing (maybe barely).

```suggestion
```

- Change: Consistent toolbar sizes with dedicated NWidgetParts. (#14344)

> - Change: Reflow Textfile window content incrementally.

Reviewer: Not player-facing (?)

```suggestion
```

> - Change: Don't measure group name widths in company livery window. (#14291)

Reviewer: Not player-facing.

```suggestion
```

- Change: [Script] Don't allow scripts to instantiate ScriptEvent (#14276)
- Change: [Script] ScriptVehicleList_Station accepts an optional VehicleType parameter (#14260)
- Change: Remove Apply button from NewGRF config window when unneeded. (#14254)
- Change: Remove extra close buttons from some windows. (#14124)
- Change: Move OpenTTD version to intro viewport. (#14233)
- Change: [NewGRF] Increase the textstack for all callbacks to 16 registers.
- Change: Rearrange buttons in the intro screen, remove unnecessary ones.

> - Change: Remove the limit of 20 parameters to a Script Text string. (#14193)

Reviewer: ?

```suggestion
- Change: [Script] Remove the limit of 20 parameters to a Script Text string. (#14193)
```

> - Change: Use same padding as NewGRF parameters for AI/GS parameters lists. (#14118)

Reviewer: Tiny UI tweak, not very noticeable by players.

```suggestion
```

> - Change: ParseInteger now checks for trailing junk chars.

Reviewer: Part of C-to-CPP changes (#14160), not player-facing.

```suggestion
```

> - Change: [NewGRF] If Action123 does not resolve in a valid SpriteSet, prefer drawing the default sprite instead of an invalid sprite.

Reviewer: Might be `Fix:` ?

```suggestion
- Change: [NewGRF] If Action123 does not resolve in a valid SpriteSet, prefer drawing the default sprite instead of an invalid sprite. (#14143)
```

> - Change: Draw info headers as shade of window colour instead of pixel colour. (#14123)

Reviewer: Only player-facing for the NewGRF window info panel.

```suggestion
```

- Change: [NewGRF] Animation-trigger 'construction stage changed' of houses and industries now also triggers at construction start. (#14089)
- Change: [Script] Reject scripts using negative version (#14096)
- Change: [NewGRF] Provide shared random bits in multi-tile animation-triggers of airport tiles and objects, just like for other features. (#14090)
- Change: Move Game Options toggles to before label, add Off/On text. (#14097)
- Change: Remove minimum width from advanced settings panel of Game Options window. (#14088)

> - Change: Use standard mauve/yellow scheme for game options window.
> - Change: Adjust section names in Graphics options tab.
> - Change: Merge Game Options and Game Settings together.

Reviewer: ```suggestion
- Change: Merge Game Options and Game Settings together. (#13242)
```

- Change: Enhance visibility of disabled bool button. (#14071)
- Change: Use scaled instead of fixed dimensions for framerate graphs. (#14039)
- Change: Include dragged train in depot tile length display. (#14060)
- Change: Use standard boolean toggle buttons for game options toggles.

> - Change: Draw boolean toggle as a slider widget.

Reviewer: ```suggestion
- Change: Use standard boolean toggle buttons for game options toggles. (#14051)
```

- Change: "setting" console command now shows default value (#14061)
- Change: Allow Double-Ctrl+Click on default size box to clear saved size. (#14055)
- Change: Don't select content when toggling its download status. (#14059)

> - Change: Add scrollbar to infrastructure window. (#14056)

Reviewer: Players have been asking for this for a while, let's make it Add (or maybe even Feature 😛 ) for visibility.

```suggestion
- Add: Add scrollbar to infrastructure window. (#14056)
```

> - Fix c2d4098afa: Crash when accessing unconfigurable badge feature. (#14403)

Reviewer: Bug appeared post-beta2.

```suggestion
```

- Fix: Include sort button width in content list header minimal size. (#14402)
- Fix: Don't draw truncation ellipsis if it's too wide. (#14401)
- Fix: Don't include ellipsis width in RTL truncation offset. (#14400)
- Fix 319caef58f: Buildings with non-zero subtile offsets broken in house picker. (#14390)

> - Fix #14361: Removing a badge filter did not clear it. (#14379)

Reviewer: Bug appeared post-beta2.

```suggestion
```

- Fix #14081: Check if removed item is a savegame (#14371)
- Fix: Allow object picker to resize if there are no classes. (#14372)

> - Fix: Mis-sized widgets due to missing widget fill. (#14370)
> - Fix: Don't include height of blank text in frame and inset widgets.

Reviewer: Not player-facing.

```suggestion
```

- Fix: Train path reservations on different railtypes could join leading to train crashes. (#14366)
- Fix #14362, 4b677e8256: Class with non static methods may have a default constructor (#14363)
- Fix #14356, ff7eb996e6: Incorrect sprite group chosen for stations and road stops. (#14359)
- Fix 0455627d16: Order backups were no longer created when selling vehicles. (#14352)

> - Fix 20738865f0: Remove colour codes from Game Options strings. (#14348)

Reviewer: Bug appeared post-beta2.

```suggestion
```

- Fix: [Script] Make ScriptOrder functions aware of road waypoints (#13419)
- Fix: Wrong error message about script Save() returned value (#14334)

> - Fix #14333, 20d83677eb: Incorrect change from index to range iteration. (#14335)

Reviewer: Bug appeared post-beta2.

```suggestion
```

- Fix: Drag drop line position when dragging NewGRF from file to active panel (#14316)
- Fix 4c8f1b0f81: Hidden-by-facility station signs were not ignored when handling clicks. (#14326)
- Fix: [Script] SQOpsLimiter had no effect for native valuator/filter (#14322)
- Fix bd1a3fe0b7: Crash if loading compatibility scripts fails. (#14318)
- Fix 5664b1e2f6: Upgrade button in NewGRF window no longer worked. (#14315)
- Fix #14295, 0455627d16: v->cur_implicit_order_index is never INVALID_VEH_ORDER_ID (#14296)
- Fix 0455627d16: Incorrect script order position mapping. (#14294)
- Fix #14234, 5aa4ad5135: Crash due to unexpected NUL char in string. (#14235)
- Fix: Clear rail vehicle flipped flag if reverse probability callback returns false. (#14281)
- Fix #14271, 7fd0e6c27d: Crash due to incorrect parameter index in string. (#14272)

> - Fix: Restore the behaviour when entering numbers in query windows: clamp integers out of range to the maximum valid value.

Reviewer: ```suggestion
- Fix: Restore the behaviour when entering numbers in query windows: clamp integers out of range to the maximum valid value. (#14268)
```

- Fix b2c57123: [Script] Conversion from ScriptEvent to ScriptEventCompanyTown subclasses (#14274)
- Fix f8bdc1e612: Empty first parameter of EncodedString was skipped. (#14273)
- Fix #14256, dfd9fbf873: Company finance windows not updated when paused. (#14258)
- Fix 0829604b4f: Dragging of selected order was still not possible. (#14259)
- Fix 0ce30d05c8: Replacing a dropdown list could reposition it partially off-screen. (#14261)

> - Fix 3cf9b159: Incorrect tile colour in smallmap for snowy MP_CLEAR tiles in vegetation mode

Reviewer: ```suggestion
- Fix 3cf9b159: Incorrect tile colour in smallmap for snowy MP_CLEAR tiles in vegetation mode (#14257)
```

- Fix #14251: Out-of-bounds writes via TextRefStack. (#14255)
- Fix 51a7edd: NewGRF debug sprite picker was broken. (#14246)
- Fix #14241, 6e10584: Invalid string parameters in subsidies list window (#14243)
- Fix #13854: 40bpp-anim blitter recolouring failed for 32bpp pixels without mask channel. (#14242)
- Fix #14234, 16b4e73: Skip PLURAL data if parameter is invalid. (#14236)
- Fix #14216, d030d17: RealSpriteGroups referencing CallbackResultSpriteGroups were always treated as callback-failure. (#14218)

> - Fix: Closing the Game Options window closes all textfile windows.

Reviewer: ```suggestion
- Fix: Closing the Game Options window closes all textfile windows. (#14210)
```

- Fix: Zoom-scroll extra viewports only if the mouse cursor is over the viewport. (#14209)
- Fix: Changing monospace font didn't refresh the monospace width cache. (#14185)
- Fix #14098: Use 64-bit type for MIDI realtime informatio [to allow longer tracks to play]. (#14144)
- Fix: Crash/Undefined behaviour in station view window. (#14183)
- Fix: Small order list arrow was wrong direction for RTL text. (#14174)
- Fix: [Script] Access to enum/consts defined outside of main.nut (#14176)
- Fix: Handle \t when loading GS strings from savegame (#14180)
- Fix bf8a241f: Handle \t in trailing whitespaces (#14178)
- Fix 66733e2a: Textbuf was broken, and did not insert any input. (#14173)
- Fix #14170, 7fd0e6c27d: Missing spaces for timetabled order strings. (#14172)
- Fix 261e09a549: Compilation failure with Fluidsynth enabled. (#14167)
- Fix #14166, dfd9fbf873: Loan was no longer invalidated when refreshing finance window. (#14168)

> - Fix: [NewGRF] Computed VarAction2 callback results were not always properly masked to 15 bit results.

Reviewer: ```suggestion
- Fix: [NewGRF] Computed VarAction2 callback results were not always properly masked to 15 bit results. (#14138)
```

- Fix: [NewGRF] The result of Action123 evaluation affected rerandomisation in a weird corner case. (#14139)

> - Fix: [NewGRF] AI station construction callback did not work for stations with ID >= 0x100.

Reviewer: ```suggestion
- Fix: [NewGRF] AI station construction callback did not work for stations with ID >= 0x100. (#14142)
```

> - Fix: Out-of-bounds read, if NewGRF stations provided no spritesets.

Reviewer: ```suggestion
- Fix: Out-of-bounds read, if NewGRF stations provided no spritesets. (#14140)
```

- Fix: List Ctrl+Click behaviour for vehicle details tooltip (#14127)

> - Fix 29ceaf0a8: checking std::optional against nullptr caused crash

Reviewer: ```suggestion
- Fix 29ceaf0a8: checking std::optional against nullptr caused crash (#14133)
```

- Fix #14125, 828e808444: NewGRF sounds were not loaded. (#14130)
- Fix #14104, 67a0fcc: Don't base timetable timing decisions on client settings. (#14105)
- Fix 5575379: Crash with default airports. (#14120)

> - Fix: Road vehicle blocking and train collissions used slightly inconsistent distance thresholds.

Reviewer: ```suggestion
- Fix: Road vehicle blocking and train collisions used slightly inconsistent distance thresholds. (#14114)
```

- Fix: [NewGRF] Cargo-types for airport-tile animation-triggers were not properly translated. (#14092)
- Fix: [NewGRF] For animation-triggers which do not supply a cargo-type in var18, the var18 bits should remain empty. (#14091)
- Fix #14107: Mark the company value in the company view window dirty when it changes. (#14112)
- Fix #14107: Invalidate all shared vehicle windows when changing group. (#14110)
- Fix: Focus settings filter box (only) when the tab becomes active. (#14100)
- Fix 2b80812922: Map size column in server list may be too wide. (#14094)

> - Fix: [NewGRF] Randomisation-trigger 'path reservation' did not work for waypoints.

Reviewer: ```suggestion
- Fix: [NewGRF] Randomisation-trigger 'path reservation' did not work for waypoints. (#14076)
```

- Fix a4dc7249ee: Changing language or interface scale could hang. (#14087)
- Fix: [NewGRF] For roadstop multi-tile-animation-triggers, the upper 16 bits were not the same for all tiles. (#14084)
- Fix f5ffd47: Don't skip separator while reading a record (#14083)
- Fix c7d6934816: Town view caption lost "(City)" flag. (#14082)
- Fix e114ed357d: Incorrect test of object flags allowed HQ to be removed. (#14062)
- Fix: Odd drawing and crash if scrollbar is not tall enough. (#14052)
- Fix #14044: Negative string parameters from GS were rendered as zero. (#14049)
- Fix: Incorrect string display for autoreplace wagon removal status. (#14038)
- Fix e49150f9ca: House picker crashes if selection is not available. (#14030)
- Fix: [NewGRF] Strange things happened, when using the synchronised tile loop animation trigger for houses. (#14011)
- Fix: [NewGRF] PrepareLayout worked on a copy of the data resulting in bad sprite layouts. (#14013)

> - Fix: [NewGRF] Animation speed properties of houses and stations had wrong default.

Reviewer: ```suggestion
- Fix: [NewGRF] Animation speed properties of houses and stations had wrong default. (#14005)
```

- Change: Draw boolean settings toggles as a slider widget (#14051, #14071)
- Change: Don't replace stripped control codes with '?' for scripts (#14028)
- Change: [Script] Move GSStation::GetOwner to GSBaseStation::GetOwner (#13406)
- Fix #14553: Parameter configuration of pre-action 14 NewGRFs did not work (#14554)
- Fix: File/directory titles not updated if language is changed (#14542)
- Fix: Missing space after old file type identifier (#14541)
- Fix: Account for both text and icon size in station waiting cargo display (#14535)
- Fix: Crash when buying out company with groups (#14534)
- Fix #10222: Off by one drawing lines of certain widths (#14520, #14522, #14523)
- Fix: [Script] Invalid title on GSGoal.Question windows (#14519)
- Fix: 'Map edges' GUI buttons shouldn't initialize with water on northeast edge (#14514)
- Fix #14415: Update survey option text when changing setting (#14487)
- Fix #14480: Music player playlist buttons are clickable but non-operational in intro menu (#14482)
- Fix: Missing button beeps (#14470)
- Fix #14464: Invalid string parameter in scenario editor when unable to build industry (#14465)

> - Fix: [Linkgraph] Incorrect NodeID to StationID conversion for EraseFlows (#14459)

Reviewer 2: this feels a bit internal, would appreciate any alternative suggestions

Reviewer 3: `-Fix: [Linkgraph] Incorrect management of flows causing subtle inconsistencies (#14459)`

As far as I can see/understand it, this could mess up the data in the logic after the call to `EraseFlows` which does some processing. But I haven't looked into too much detail, just that typing-wise things were incorrect.

> - Fix: [Linkgraph] Incorrect NodeID to StationID conversion for EraseFlows (#14459)

Reviewer: ```suggestion
- Fix: [Linkgraph] Use correct station ID when erasing flows (#14459)
```

- Fix: Vehicle liveries did not update when switching company (#14456)

> - Fix: EngineImageType mismatch between sizing and drawing in preview window (#14455)

Reviewer 2: this feels a bit internal, would appreciate any alternative suggestions

Reviewer 3: `- Fix: Engine preview window could have the wrong size (#14455)`

> - Fix: EngineImageType mismatch between sizing and drawing in preview window (#14455)

Reviewer: ```suggestion
- Fix: Use correct window size when vehicle preview and purchase sprites are different (#14455)
```

- Change: [NewGRF] Provide shared random bits in multi-tile animation-triggers of airport tiles and objects, just like for other features. (#14090)
- Change: Move Game Options toggles to before label, add Off/On text. (#14097)
- Change: Remove minimum width from advanced settings panel of Game Options window. (#14088)
- Change: Use standard mauve/yellow scheme for game options window.
- Change: Adjust section names in Graphics options tab.
- Change: Merge Game Options and Game Settings together.
- Change: Enhance visibility of disabled bool button. (#14071)
- Change: Use scaled instead of fixed dimensions for framerate graphs. (#14039)
- Change: Include dragged train in depot tile length display. (#14060)
- Change: Use standard boolean toggle buttons for game options toggles.
- Change: Draw boolean toggle as a slider widget.
- Change: "setting" console command now shows default value (#14061)
- Change: Allow Double-Ctrl+Click on default size box to clear saved size. (#14055)
- Change: Don't select content when toggling its download status. (#14059)
- Change: Add scrollbar to infrastructure window. (#14056)
- Fix #14385: [Strgen] Don't count parameters more than one time (#14413)

> - Fix: Inconsistent capitalization in game settings (#14412)

Reviewer 3: I wouldn't include this in the changelog, as it's basically some tweaking of a string like translators do all the time.
